iib/common/tracing.py

```diff
@@ -0,0 +1,163 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+"""Configures the Global Tracer Provider and exports the traces to the OpenTelemetry Collector.
+
+The OpenTelemetry Collector is configured to receive traces via OTLP over HTTP.
+The OTLP exporter is configured to use the environment variables defined in the ansible playbook.
+
+Usage:
+    @instrument_tracing()
+    def func():
+        pass
+
+"""
+import os
+import functools
+import logging
+from typing import Dict
+from opentelemetry import trace
+from opentelemetry.trace import SpanKind, Status, StatusCode
+from opentelemetry.sdk.resources import Resource, SERVICE_NAME
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import (
+    BatchSpanProcessor,
+)
+from opentelemetry.propagate import set_global_textmap
+from opentelemetry.trace.propagation import (
+    set_span_in_context,
+)
+from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
+from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+
+log = logging.getLogger(__name__)
+propagator = TraceContextTextMapPropagator()
+
+
+class TracingWrapper:
+    """Wrapper class that will wrap all methods of calls with the instrument_tracing decorator."""
+
+    __instance = None
+
+    def __new__(cls):
+        """Create a new instance if one does not exist."""
+        if TracingWrapper.__instance is None:
+            log.info('Creating TracingWrapper instance')
+            cls.__instance = super().__new__(cls)
+            otlp_exporter = OTLPSpanExporter(
+                endpoint=f"{os.getenv('OTEL_EXPORTER_OTLP_ENDPOINT')}/v1/traces",
+            )
+            cls.provider = TracerProvider(
+                resource=Resource.create({SERVICE_NAME: os.getenv('OTEL_SERVICE_NAME') or 'iib'})
```

Comment:
May I ask about the reason to include `or 'iib'` while creating the resource? Also, if this is a required variable, wouldn't it be better to check if it has a value other than None? I am not sure what is the required variable from the environment variables or are these all optional variables?

Reply:
You don't want to fail because this parameter is not defined. In general, it is recommended to set that value.